File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] arr) {
        int[] dp = new int[arr.length];
        dp[0] = 1;
        for (int i = 1; i < arr.length; i++) {
            int m = 0;
            for (int j = 0; j < i; j++) {
                if (arr[i] > arr[j]) {
                    m = Math.max(m, dp[j]);
                }
            }
            dp[i] = m + 1;
        }
        int maxValue = 0;
        for (int i = 0; i < dp.length; i++) {
            maxValue = Math.max(maxValue, dp[i]);
        }
        return maxValue;
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution(object):
    def lengthOfLIS(self,a):
        n = len(a)
        arr = [0] * n
        arr[0] = a[0]
        curr = 0

        

        for i in range(1, n):
            ind = self.index(arr, curr, a[i])
            arr[ind] = a[i]
            curr = max(ind, curr)

        return curr + 1
    def index(self,arr, i, k):
            ans = -1
            low = 0
            high = i
            while low <= high:
                mid = (low + high) // 2
                if arr[mid] > k:
                    ans = mid
                    high = mid - 1
                elif arr[mid] < k:
                    low = mid + 1
                else:
                    ans = mid
                    break
            if ans == -1:
                return i + 1
            else:
                return ans
        